[EXP][CMDBUF] L0 Immediate command-list support

mfrancepillois opened the pull request:
Adds support for L0 immediate command-list.
The command-list containing the graph operations is still batch command-list but graphs can now be submitted even though immediate queue is requested by users. 
Prefix and Suffix additional command list types follow Queue type.

EwanC commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -737,18 +731,29 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
+  // directly using the Level-Zero API to avoid type mismatches if using UR
+  // functions.
+  auto ZeCommandList = CommandListPtr->first;
+  auto ZeResult = ZE_CALL_NOCHECK(
```
Doesn't look like we're checking `ZeResult`, no sure if `ZE2UR_CALL` would be better here.

mfrancepillois replied:
Nice catch. Thanks :+1:

EwanC commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -759,21 +764,14 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
     ZE2UR_CALL(zeCommandListAppendBarrier,
                (SignalCommandList->first, RetEvent->ZeEvent, 1,
                 &(CommandBuffer->SignalEvent->ZeEvent)));
+    Queue->executeCommandList(SignalCommandList, false, false);
```
Think we can move `ur_command_list_ptr_t SignalCommandList{};` declaration to inside the `if (Event)` scope now

EwanC commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -737,18 +731,29 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
+                 (WaitCommandList->first, CommandBuffer->WaitEvent->ZeEvent,
+                  CommandBuffer->WaitEvent->WaitList.Length,
+                  CommandBuffer->WaitEvent->WaitList.ZeEventList));
+      Queue->executeCommandList(WaitCommandList, false, false);
```
Think we can move `ur_command_list_ptr_t WaitCommandList{};` declaration to inside this if scope now, can 
also maybe only do ` UR_CALL(Queue->Context->getAvailableCommandList(Queue, WaitCommandList,
false, false))` if wait list isn't empty

EwanC replied:
The scope i was hoping we could move `ur_command_list_ptr_t WaitCommandList{` and 

```
 UR_CALL(Queue->Context->getAvailableCommandList(Queue, WaitCommandList,
                                                    false, false))
```
to is inside this `if (!CommandBuffer->WaitEvent->WaitList.isEmpty()) {`. So we only find an available command-list if we're going to execute it.

mfrancepillois replied:
Yes. Sorry. You are right. I fixed that.

Bensuo commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -737,43 +731,47 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
+  // directly using the Level-Zero API to avoid type mismatches if using UR
+  // functions.
+  auto ZeCommandList = CommandListPtr->first;
+  ZE2UR_CALL(
+      zeCommandQueueExecuteCommandLists,
+      (ZeCommandQueue, 1, &ZeCommandList, CommandListPtr->second.ZeFence));
```
I think if we're now submitting the command list ourselves rather than through the queue we could probably not store it it in this iterator type to clean things up a little bit, since that was only necessary for that.

mfrancepillois replied:
The iterator still help us keep track of fences to release them in the command-buffer destructor. But it is true that we can also use a simpler container for that.

mfrancepillois replied:
I changed the way we track fences by using a vector instead of the CommandList map.

Bensuo reviewed (approved):
LGTM

EwanC reviewed (approved):
Looks good 👍

reble commented on source/adapters/level_zero/command_buffer.hpp:
```diff
@@ -46,6 +46,10 @@ struct ur_exp_command_buffer_handle_t_ : public _ur_object {
   ze_command_list_handle_t ZeCommandList;
   // Level Zero command list descriptor
   ZeStruct<ze_command_list_desc_t> ZeCommandListDesc;
+  // List of Level Zero fences created when submiting graph.
```
```suggestion
  // List of Level Zero fences created when submitting a graph.
```

reble reviewed (approved):
LGTM!

julianmi commented on source/adapters/level_zero/command_buffer.hpp:
```diff
@@ -46,6 +46,10 @@ struct ur_exp_command_buffer_handle_t_ : public _ur_object {
   // Level Zero command list descriptor
   ZeStruct<ze_command_list_desc_t> ZeCommandListDesc;
+  // List of Level Zero fences created when submiting graph.
+  // This list is needed to release all the Fences retained by the
```
```suggestion
  // This list is needed to release all fences retained by the
```

Commit:
[EXP][CMDBUF] L0 Immediate command-list support

Adds support for L0 immediate command-list.
The command-list containing the graph operations is still batch cmd-list but graphs can now be submitted even though immediate queue is requested by users.
Prefix and Suffix additional command list types follow the Queue type.

mfrancepillois commented:
Upstream UR PR : https://github.com/oneapi-src/unified-runtime/pull/1218
